sacc2/sacc.py

```diff
@@ -0,0 +1,1026 @@
+import numpy as np
+import copy
+import os
+import yaml
+import warnings
+import pickle
+
+from collections import OrderedDict
+from astropy.io import fits
+from astropy.table import Table, Column
+
+from .tracers import BaseTracer
+from .windows import BaseWindow
+from .covariance import BaseCovariance
+from .utils import unique_list
+
+
+# add more as we develop them
+allowed_types = [
```

Comment:
I don't see why we need to specify `galaxy`, `shear` or `ggl`. These are just different combinations of spin-0 and spin-2 two-point functions. E.g. why should we add another type when we want to incorporate CMB lensing x delta_gal correlations? That should be encompassed by what currently is `galaxy_density_cl` or `galaxy_density_w`, i.e. the correlation of two spin-0 fields.

Reply:
Yes, I'm not 100% sure about this.  We could store things just in terms of the two fields, as we did in the 2pt code.  But there are some combinations like magnification which can be hard to distinguish from others if you just specify the two tracers and their types.  So I'm really not sure.  Also if you switch to a different statistic like mass aperture then you need a new name.

Reply:
Can you point me to one example of a magnification data vector, just so I understand this better?
The way I thought about this, at least in the context of an N-point function, is that the tracers tell you the quantity that you're correlating (number counts, magnification, shear, whatever), while the type tells you the combination of spin components you're correlating and whether it's in real or Fourier space. I think this should be enough to specify everything, but I may be missing something.

Since a user can always define a new tracer with some degree of freedom, this is then easy to extend to any quantity we may care about in cosmology without having to add a bunch of new types every time we want to include a new observable (since in general we will want to include all its possible correlations with all existing observables).

Reply:
I was just meaning that if you do mag x mag then it's another spin-0 galaxy correlation, just like counts x counts.  But you're right we can include another piece of info about what it is.

What about things like COSEBIs that are combinations of other 2pt functions?

But as I say, I'm really not sure what the best way of doing this is!

Reply:
In original sacc we have an additional concept of a physical tracer, i.e.  a set of galaxies, from which you can derive shear or number counts, but in retrospect I agree that that was a bad idea (I had some vague notion about correlated systematics, but now I think there are probably better ways of doing that). So, once tracer = one field on the sky. But quantity is a separate thing, not only there are real and config, there are also various compensated statistc, etc as you say.

Reply:
So I've tried to summarize this in my head, and have come to these two options:

data_type is a string
- pro: less to pass around when creating (slightly simpler)
- pro: slightly faster searches (not a big difference)
- con: need to create new ones for new combinations, more verbose


data_type is a pair of strings
- pro: don't have to create a new name when a new type of data is added
- pro: potentially more explicit about what things are
- pro: perhaps easier to automate code that uses the two individually
- con: non-2pt data does not fit this model at all! e.g. cluster things.

To me the last of these is the most important - if we want this to work for non-2pt data then we need to keep the data type as a string, at the risk of increasing the number of combination unnecessarily for 2pt data.

Reply:
Have I missed something?  If not I'll add any that we can think of now and put it into a separate python file so it's easier to add things later.

Reply:
It has to be a string - my intention is to use this format for everything

Reply:
Wait, but data_type in this case would be `number_counts` or `one_point` or something like that. It'd be associated to a single tracer, corresponding to the number of clusters in a given redshift-richness bin.

This is the way we thought about structuring things when we discussed this particular problem with Elisabeth and Hironao.

Reply:
This is also good in the sense that it allows you to include all two-point functions between any observable and a particular cluster bin (you just reference the corresponding cluster `tracer`).

Reply:
Not following @damonge. Note that I want to use this format for MCMC chains too. I mean literally everything.

Reply:
The data_type string should let you know what calculation you need to do to predict the value.  The tracers and window (etc)  then contain metadata for that calculation.

This is *definitely* not designed for MCMC chains!  They are always homogenous, so much more suited to existing formats.

Reply:
The way we thought about this: the data_type tell you about the maths of the data point: whether it's an N-point function, what spin components, Fourier or real. The tracers that each data point is associated with then tell you everything you need to know about the physical quantities involved. You need both pieces of information in order to compute the prediction.

Structuring things this way is good, I think, because it compartimentalises things better, and it prevents you from defining N^D data types if you are considering N different tracers and up to D-point functions. You only need to define N tracers, and those can actually be fully defined by the user.

Reply:
Yes, please, let's not use this for MCMC, catalogs, images, TODs or anything else ;-)

Reply:
The maths of the data point isn't enough - it needs to tell us the physics as well.
For example, CMB shear fields and galaxy shear fields have the same mathematical structure, but we're feeding this data into firecrown, which needs to know whether to use a delta function n(z) or a distribution, which changes how you do the integral (or don't do it!) completely.  So the data_type needs to tell us what entire calculation you have to do, so that means it should be something like galaxy_shear_ee rather than just "2pt_spin2_ee" or something like that.

Reply:
Designed vs used is a very different statement @joezuntz :p

Reply:
I don't think so. In this particular case the data_type would be `ClEE`, and the two tracers would be a shear tracer (identified by a tag in it that says it's shear, and containing an N(z)), and a CMB lensing tracer (identified by a tag in it that says it's CMB lensing, and not much else).

The tracers just need to have tags that identify them, and those tags don't have to be standardized (i.e. users could create new tracers if they need them). Whatever code reads the file just needs to be able to interpret these tags (firecrown in this case).

Reply:
I think we just disagree here - I'd like this to be more explicit and for someone looking at the data to be able to tell explicitly at a glance at the row what kind of data it is. And you'd prefer something defined more mathematically and with minimal repetition by encoding this information in the tracer.

Anyone else have any thoughts?

Reply:
yes, I think that sums it up ;-)

In my favour, I think this setup would also make your life easier in firecrown: for a given datapoint, the tracers tell you directly what pair of CCL objects you need to correlate. Otherwise you'll need to write a dictionary that maps datatypes to pairs of tracers.

Reply:
Since the initial discussion about cluster implementation (which seems to be the outlier here) was initiated with @elikrause , it might be worth pinging her.

Reply:
For what it is worth, I tend to agree with David here. At the end of the day, we'll have manually marginalise over all kinds of crap, so firecrown will likely do things differently for different galaxy shear, i.e. you don't want to have a special data type for red and blue galaxies, just because intrinsic alignments is different. So let this field just describe the quantity and the recipe for calculation comes from the particular tracer...

Reply:
@beckermr your thoughts here would be useful too.

Reply:
Okay, I think I'm just going to make a decision here and stick with what I've done so far, as ultimately it's not that important either way, and I appointed myself the sacc champion at Becker's demand.   Does anyone have any more comments?  I will document the code better and ask for a code review after that.

Reply:
I think this will make things difficult when it comes to bringing in more probes or when using this for other experiments. Then again, since I haven't done the work, I don't want to block any progress.

That being said, I will feel free to submit a PR with what I think is be the best way of doing this later on ;-)

Reply:
Opening an issue for this.